### PR TITLE
Create CrowdSignal survey views and navigation

### LINKED ISSUE #2644: [IAFeedback] Show Crowdsignal Survey
The #2641 task should be completed first. 

Screen Shot 2020-07-22 at 7 08 36 PM

Pressing on the **Could Be Better** button should show the **generic** Crowdsignal survey. 

See p91TBi-2Sy-p2 for details. We can use a temporary Crowdsignal survey URL for now. We can replace it later. 

## Reusability

The components we create for the Crowdsignal survey will be reused in #2643. 

## Others

- [ ]  Must be hidden behind the feature flag

### PR DESCRIPTION
**Pull request by ThomazFB**
Summary
==========
Fixes issue #2644, this PR delivers the following features:
- Trigger Crowdsignal Feedback Survey when `Could be better` button is clicked on the `FeedbackRequestCard`
- Crowdsignal Feedback Survey view using a WebView
- WebView lifecycle handling with loading until survey view is completely loaded and survey completion interception
- Feedback Completed view with Dark Mode support
- Feedback fragments navigation, back stack and exit handled through Navigation Controller
- Contact Us text button integrated with `Help & Support` view

Changelog
==========
| Description | Commit |
| ------------- | ------------- |
| Initiate Survey fragments implementation | d0716b4d3d8290da28bbed92fbf200a17fb513c5 |
| Integrate Survey fragments with the Navigation Controller declaration structure |  9f3272e28df6da4572918cacf342d9570b867f64 |
| Create interception of the WebView lifecycle to make possible the loading dialog presentation and removal | e08773639d83e2425390b3b889ec247b3909386f |
| Make JavaScript execution available inside the WebView in order to better fit CrowdSignal web navigation experience | e55f021173383a6f2fb4b90b1a2cb0a78aad8ab5 |
| Create rule to intercept URL query parameter interaction through request loading override to verify if CrowdSignal survey is complete | 440f3c8fb9fb4d07829646ddf236f581cb4570b7 |
| `FeedbackCompletedFragment` created | db0dbad1b423964cde5df9d49acc2c5ee32b673d |
| `FeedbackCompletedFragment` UI structure defined | 530b3c6ac825e3766c1ab9531fdf11df5b06f786 |
| Change Survey URL declaration to the `AppUrls` object | 002f7f15062b54fa471d9f3e6d7f2f9dc103396b |
| Make the Crowdsignal Feedback Survey navigation start available from the `DashboardFragment` view | 2398d88a241251b4a1ea4933a89465445185291f |

Screenshots
==========
Light Mode
------------
| Click on `Could be better` button | Wait for the survey loading | Interact with the survey until it's finished | Verify if the expected view is presented and is fully functional | 
| ------------- | ------------- | ------------- | ------------- |
| Screenshot_1597814080 | Screenshot_1597812682 | Screenshot_1597891397 | Screenshot_1597812776 |

Dark Mode
------------
| Click on `Could be better` button | Wait for the survey loading | Interact with the survey until it's finished | Verify if the expected view is presented and is fully functional | 
| ------------- | ------------- | ------------- | ------------- |
| Screenshot_1597891515 | Screenshot_1597891046 | Screenshot_1597891102 | Screenshot_1597891109 |

Tablet
------------
| Click on `Could be better` button | Wait for the survey loading | Interact with the survey until it's finished | Verify if the expected view is presented and is fully functional | 
| ------------- | ------------- | ------------- | ------------- |
| Screenshot_1597891681 | Screenshot_1597891688 | Screenshot_1597891695 | Screenshot_1597891708 |

How to Test
------------

1. Access the `My Store` section from the Woo app
2. Inside the `FeedbackRequestCard`, click in the `Could be better` button
3. Verify that the survey view is called, the loading dialog is visible until the survey is completely loaded
4. Interact with the survey until the conclusion, after finishing it, verify that the custom native Feedback Completed View is called instead of letting the WebView show the default web one.
5. Once inside the Feedback Completed View, verify that the each button is working as expected and described

Update release notes:

- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by Garance91540 — CHANGES_REQUESTED** [review]
Hi Thomaz, looking good! 

I have a few comments/questions:
- Can you please share with me dark mode screenshots and one iPad screenshot too?
- Which survey are you using exactly? I think we should use the same on Android and iOS so we only have on place to go to see all the data. Here's the Crowdsignal survey @shiki worked on https://automattic.survey.fm/woo-app-general-feedback-test-survey, can we use it for Android too?

**Comment by ThomazFB** [comment]
Hey @Garance91540! Thanks for your inputs!
>Can you please share with me dark mode screenshots and one iPad screenshot too?

Sure! I've added into the PR description the navigation sequence running in Dark Mode and on an Android Tablet 😄 

>Which survey are you using exactly? I think we should use the same on Android and iOS so we only have on place to go to see all the data. Here's the Crowdsignal survey @shiki worked on https://automattic.survey.fm/woo-app-general-feedback-test-survey, can we use it for Android too?

I've intentionally left the PR with the test survey URL just for review purposes, but to make easier the design review I've just added the production URL with the CSS changes made by @shiki. Thanks for pointing that out!

Let me know if these changes works for you!

**Review by Garance91540 — DISMISSED** [review]
Looking awesome, thank you @ThomazFB !

**Comment by shiki** [comment]
Regarding the style, looks like the radio buttons are squares instead of circles. 😄 I'll fix those next sprint. (https://github.com/woocommerce/woocommerce-ios/issues/2546)

**Review by AmandaRiu — CHANGES_REQUESTED** [review]
@ThomazFB Nice work on this 🥳 Here's some feedback from what I found during testing:

- I noticed that if the device is rotated the survey loses it's values. Is there a way we can prevent that from happening?
- The "Feedback sent" screen does not scroll so you can't see the rest of the view in landscape mode or click the "Back to Store" button:



@shiki I noticed the survey errors look like this on Android, is this expected?

**Comment by shiki** [comment]
@AmandaRiu Yes, it's a problem in Crowdsignal's CSS editor. I reported it to them (p91TBi-2RL-p2#comment-1530). Thank you for confirming. 🙂

**Comment by ThomazFB** [comment]
Hey @AmandaRiu! Thank you so much for your review, I've submitted some change and I think the code is ready for a second round!

>I noticed that if the device is rotated the survey loses it's values. Is there a way we can prevent that from happening?

Thank you for pointing that out! I did some investigations regarding WebView state management and it doesn't look very reliable, nevertheless, I did some additional changes and restored my previous attempts to restore the WebView previous state and it looks to be working now. Let me know how that goes for you. Here's the changes: https://github.com/woocommerce/woocommerce-android/pull/2723/commits/1034133023f225423111c47cd92ea72f01c965a0

>The "Feedback sent" screen does not scroll so you can't see the rest of the view in landscape mode or click the "Back to Store" button

Sorry about that! Here's the change allowing scrolls on the UI: https://github.com/woocommerce/woocommerce-android/pull/2723/commits/7803c858618345e84f8adc5d2276f7189875fcc9

File: WooCommerce/src/main/res/layout/fragment_feedback_completed.xml [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <androidx.constraintlayout.widget.ConstraintLayout
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:gravity="center_horizontal|top"
+        android:orientation="vertical"
+        android:padding="32dp">
```
Comment (AmandaRiu) [inline WooCommerce/src/main/res/layout/fragment_feedback_completed.xml]: np: hard code values. Should use one of our established values in `dimens.xml` - same for the rest of this file.

File: WooCommerce/src/main/res/layout/fragment_feedback_completed.xml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+        <TextView
```
Comment (AmandaRiu) [inline WooCommerce/src/main/res/layout/fragment_feedback_completed.xml]: should use a `MaterialTextView`

File: WooCommerce/src/main/res/layout/fragment_feedback_completed.xml [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            android:id="@+id/completion_title"
+            style="@style/Woo.TextView.Subtitle1"
+            android:layout_width="241dp"
+            android:layout_height="48dp"
```
Comment (AmandaRiu) [inline WooCommerce/src/main/res/layout/fragment_feedback_completed.xml]: Why hard-coded widths and heights?

**Review by AmandaRiu — CHANGES_REQUESTED** [review]
@ThomazFB The scrolling thing is fixed 👍 thanks for taking care of that. I'm still seeing some issues when rotating the survey. The email disappears:

ezgif com-video-to-gif

Other than that I left a couple smaller comments in code. It's really close to merging! 👍

File: WooCommerce/src/main/res/layout/fragment_feedback_completed.xml (hunk 3 above)
Comment (ThomazFB) [reply WooCommerce/src/main/res/layout/fragment_feedback_completed.xml]: My mistake here, sorry about that, I've fixed this for all the missing extraction dimensions right here: https://github.com/woocommerce/woocommerce-android/pull/2723/commits/c1ad309bd828fa668d181fb13f41ce708841a6b1

File: WooCommerce/src/main/res/layout/fragment_feedback_completed.xml (hunk 2 above)
Comment (ThomazFB) [reply WooCommerce/src/main/res/layout/fragment_feedback_completed.xml]: Oops, forgot to adjust that one, it's fixed right here: https://github.com/woocommerce/woocommerce-android/pull/2723/commits/c1ad309bd828fa668d181fb13f41ce708841a6b1

File: WooCommerce/src/main/res/layout/fragment_feedback_completed.xml (hunk 1 above)
Comment (ThomazFB) [reply WooCommerce/src/main/res/layout/fragment_feedback_completed.xml]: Sure! Thanks for that! It's done right here: https://github.com/woocommerce/woocommerce-android/pull/2723/commits/c1ad309bd828fa668d181fb13f41ce708841a6b1

**Comment by ThomazFB** [comment]
Hey @AmandaRiu, thank you again for your review. I've pushed all the missing dimension extractions on the `xml` files. But there's one thing I'm worried about the `WebView` state handling. I did a lot of attempts trying to find a reliable way to restore the state on the `email` field, but it looks like this field uses JS to control their state because of pattern matching to verify if it's a valid e-mail address, and since the Android WebView instance state storing bundle doesn't cover JS variable states and logic control, we see this kind of behavior happening.

Because of that same characteristic of the survey webpage, if we access the [Survey link](https://automattic.survey.fm/woo-app-general-feedback-test-survey) from a desktop browser like Chrome, we can see that even in that case the browser has a hard time to keep the email state in scenarios like pressing back and forward buttons on the browser tab.

I think we'll need to assume some unwanted compromises in order to achieve a good state handling, like blocking the Activity from recreating itself when we do a configuration change like rotation, but I think that's a way too harsh measure since it will affect a lot of other Fragments and Views, and overall it's a not good practice to be adopted. Not sure if we have many options to solve this problem. WDYT?

I'm curious if the iOS part managed to solve this problem in some way. CC: @shiki @Ecarrion

**Comment by Ecarrion** [comment]
@ThomazFB 
> I'm curious if the iOS part managed to solve this problem in some way.

I just tested it and it doesn't seem to be a problem iOS side, I assume it's because iOS doesn't recreate view upon rotation, just re-layout them.

**Comment by AmandaRiu** [comment]
> I think we'll need to assume some unwanted compromises in order to achieve a good state handling, like blocking the Activity from recreating itself when we do a configuration change like rotation, but I think that's a way too harsh measure since it will affect a lot of other Fragments and Views, and overall it's a not good practice to be adopted. Not sure if we have many options to solve this problem. WDYT?

I agree that preventing a configuration change on rotation would be too harsh for something like this. I think we should leave it for now. The good news is that the survey will not be submitted without the email form filled so the user will be asked to re-enter it if a configuration change happens. And at least the rest of the survey is 👍

**Review by AmandaRiu — APPROVED** [review]
LGTM! :shipit:

**Pull request merged**